cpp/oneapi/dal/backend/primitives/reduction/reduction_rm_cw_wrapper_dpc.cpp

```diff
@@ -0,0 +1,120 @@
+/*******************************************************************************
+* Copyright 2021 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "oneapi/dal/backend/primitives/reduction/reduction_rm_cw_dpc.hpp"
+
+#include "oneapi/dal/backend/primitives/common.hpp"
+
+namespace oneapi::dal::backend::primitives {
+
+template <typename Float, typename BinaryOp, typename UnaryOp>
+reduction_rm_cw<Float, BinaryOp, UnaryOp>::reduction_rm_cw(sycl::queue& q) : q_{ q } {}
+
+template <typename Float, typename BinaryOp, typename UnaryOp>
+typename reduction_rm_cw<Float, BinaryOp, UnaryOp>::reduction_method
+reduction_rm_cw<Float, BinaryOp, UnaryOp>::propose_method(std::int64_t width,
+                                                          std::int64_t height) const {
+    if (height >= device_max_wg_size(q_) && height > width) {
+        return reduction_method::naive_local;
+    }
+    return reduction_method::naive;
+}
+
+template <typename Float, typename BinaryOp, typename UnaryOp>
+sycl::event reduction_rm_cw<Float, BinaryOp, UnaryOp>::operator()(
+    const typename reduction_rm_cw<Float, BinaryOp, UnaryOp>::reduction_method method,
```

Comment:
Why `typename reduction_rm_cw<Float, BinaryOp, UnaryOp>::reduction_method` is needed? Just use `reduction_method `. If you are concerned about visibility of type `reduction_method` outside of class declaration, try to compile this code:
```c++
template <typename T>
class A {
private:
    using my_type = T;
public:
    void foo(my_type x);
    my_type bar();
};

template <typename T>
void A<T>::foo(my_type x) {}

// Don't do that!
// template <typename T>
// void A<T>::foo(typename A<T>::my_type x) {}

template <typename T>
auto A<T>::bar() -> my_type {
    return my_type(0);
}

// This will not compile
// template <typename T>
// my_type A<T>::bar() {
//     return my_type(0);
// }

// Don't do that!
// template <typename T>
// typename A<T>::my_type A<T>::bar() {
//     return my_type(0);
// }

int main() {
    A<float>{}.foo(0.0f);
    A<float>{}.bar();
}
```

Reply:
Fixed